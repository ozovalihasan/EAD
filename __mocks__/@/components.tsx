import React from "react";
import { 
  CustomHandleType, 
  HasAnyEdgePropsType, 
  MarkerBaseType, 
  MarkerType, 
  RemoveEdgeButtonType, 
  ShowEdgeTextType, 
  SidebarOptionsType, 
  ToggleOptionalButtonType
} from "../../src/components"

export const AttributeTypeOptions = () => (
  <>
    MockAttributeTypeOptions
  </>
)

export const SidebarOptions = (props: SidebarOptionsType) => (
  <>
    MockSidebarOptions
    { 
      Object.keys(props)
        .map((key) => `${key}: ${props[key as keyof typeof props]}`) 
    }
  </>
)

export const Sidebar = (
  () => (
    <>
      MockSidebar
    </>
  )
)

export const Navbar = (
  () => (
    <>
      MockNavbar
    </>
  )
)

export const SidebarHandle = (
  () => (
    <>
      MockSidebarHandle
    </>
  )
)

export const MainReactFlow = (
  () => (
    <>
      MockMainReactFlow
    </>
  )
)

export const MarkerBase = ({children, ...rest}: MarkerBaseType ) => (
  <>
    MockMarkerBase
    { Object.keys(rest).map((key) => `${key}: ${rest[key as keyof typeof rest]}`) }
    {children}
  </>
)

export const EADLogo = (
  () => (
    <>
      MockEADLogo
    </>
  )
)

export const GithubLogo = (
  ( ) => (
    <>
      MockGithubLogo
    </>
  )
)

export const Settings = (
  ( ) => (
    <>
      MockSettings
    </>
  )
)

export const CrossMarker = (
  (props: MarkerType ) => (
    <>
      MockCrossMarker
      { Object.keys(props).map((key) => `${key}: ${props[key as keyof typeof props]}`) }
    </>
  )
)

export const CrowsFootMarker = (
  (props: MarkerType ) => (
    <>
      MockCrowsFootMarker
      { Object.keys(props).map((key) => `${key}: ${props[key as keyof typeof props]}`) }
    </>
  )
)

export const CircleLineMarker = (
  (props: MarkerType ) => (
    <>
      MockCircleLineMarker
      { Object.keys(props).map((key) => `${key}: ${props[key as keyof typeof props]}`) }
    </>
  )
)

export const StraightLineMarker = (
  (props: MarkerType ) => (
    <>
      MockStraightLineMarker
      { Object.keys(props).map((key) => `${key}: ${props[key as keyof typeof props]}`) }
    </>
  )
)

export const RemoveEdgeButton = (
  (props: RemoveEdgeButtonType ) => (
    <>
      MockRemoveEdgeButton
      { Object.keys(props).map((key) => `${key}: ${props[key as keyof typeof props]}`) }
    </>
  )
)

export const ToggleOptionalButton = (
  (props: ToggleOptionalButtonType ) => (
    <>
      MockToggleOptionalButton
      { Object.keys(props).map((key) => `${key}: ${props[key as keyof typeof props]}`) }
    </>
  )
)

export const ShowEdgeText = (
  ({label, ...rest}: ShowEdgeTextType ) => ( 
    <>
      MockShowEdgeText
      { Object.keys(rest).map((key) => `${key}: ${rest[key as keyof typeof rest]}`) }
      label: { label }
    </>
  )
)


export const CustomHandle = (props: CustomHandleType) => {
  
  return (
    <>
      MockCustomHandle
      { Object.keys(props).map((key) => `${key}: ${props[key as keyof typeof props]}`) }
    </>
  );
}

export const HasAnyEdge = (
  (
    props: HasAnyEdgePropsType 
  ) => (
    <>
      MockHasAnyEdge
      { Object.keys(props).map((key) => `${key}: ${props[key as keyof typeof props]}`) }
    </>
  )
)

export const MainColorDropdown = () => (
  <div>
    MockMainColorDropdown
  </div>
)

export const HasOneHandle = ({ nodeId }: {nodeId : string}) => (
  <div>
    MockHasOneHandle
    nodeId: {nodeId}
    has_one
  </div>
)

export const HasManyHandle = ({ nodeId }: {nodeId : string}) => (
  <div>
    MockHasManyHandle
    nodeId: {nodeId}
    has_many
  </div>
)

export const ThroughHandle = ({ nodeId }: {nodeId : string}) => (
  <div>
    MockThroughHandle
    nodeId: {nodeId}
    through
  </div>
)

export const TargetHandle =(
  ({ nodeId }: {nodeId: string} ) => (
    <>
      MockTargetHandle
      nodeId: {nodeId}
    </>
  )
)

export const TableName =(
  ({ nodeId, tableId }: {nodeId: string, tableId: string} ) => (
    <>
      MockTableName
      nodeId: {nodeId}
      tableId: {tableId}
    </>
  )
)

export const AllHandlers =(
  ({ nodeId }: {nodeId: string} ) => (
    <>
      MockAllHandlers
      nodeId: {nodeId}
    </>
  )
)

export const SelectThroughNode =(
  ({ nodeId }: {nodeId: string} ) => (
    <>
      MockSelectThroughNode
      nodeId: {nodeId}
    </>
  )
)

    