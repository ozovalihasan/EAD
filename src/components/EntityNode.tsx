import { useCallback, useEffect } from 'react';
import { Handle, Position } from 'react-flow-renderer';
import {
  HasManyHandle,
  HasOneHandle,
  TableName,
  ThroughHandle 
} from "."
import useStore from '../zustand/store';


export type EntityNodeDataType = {
  tableId: string,
  name: string,
}

export const EntityNode = ({id, data, selected }: {id: string, data: EntityNodeDataType, selected: boolean}) => {
  const { onNodeInputChange, onMouseEnterThrough, onChangeAssociationType } = useStore();

  const isConnectContinue = useStore(state => state.isConnectContinue)
  const isMouseOnNode = useStore(state => state.isMouseOnNode)
  const mouseOnNodeId = useStore(state => state.mouseOnNodeId)
  const selectedNodeForThrough = useStore(state => state.selectedNodeIdForThrough) == id
  const associationType = useStore(state => state.associationType)
  const connectionStartId = useStore(state => state.connectionStartId)

  const visibleSourceHandles = ((!isConnectContinue && isMouseOnNode && mouseOnNodeId == id) ? "visible" : "hidden")
  
  return (
    <div className={`border-black border border-solid p-1 rounded-sm ${ selectedNodeForThrough ?  "bg-second-400" : "bg-first-50"} ${ selected ?  "bg-first-200" : ""}`} >
      
      <Handle className="border-none w-6 h-6" type="target" position={Position.Top} id="top" style={{visibility: isConnectContinue ? "visible" : "hidden"}}/>
      
      <div>
        <label htmlFor="text"></label>
        <input placeholder='Entity' value={data.name} className="w-32 p-1 rounded-md" id="text" name="text" onChange={(event) => onNodeInputChange(event, id)} />
      </div>
        {data && <TableName tableId={data.tableId}></TableName>}

      
      
      <div className='relative w-full'>
        <HasOneHandle visibility={visibleSourceHandles}/>
        <Handle
          onMouseDown={() => onChangeAssociationType("has_one", id)}
          className="opacity-50 absolute -bottom-3 justify-center items-center flex border-none w-6 h-6 left-1/4"
          style={{visibility: visibleSourceHandles}}
          type="source"
          position={Position.Bottom}
          id="bottom1"
        />

        <HasManyHandle visibility={visibleSourceHandles}/>
        <Handle
          onMouseDown={() => onChangeAssociationType("has_many", id)}
          className=" opacity-50 absolute -bottom-3 justify-center items-center flex border-none w-6 h-6 left-1/2"
          style={{visibility: visibleSourceHandles}}
          type="source"
          position={Position.Bottom}
          id="bottom2"
        />

        <ThroughHandle visibility={visibleSourceHandles}/>
        <Handle
            onMouseDown={() => onChangeAssociationType("through", id)}
            className=" opacity-50 absolute -bottom-3 justify-center items-center flex border-none w-6 h-6 left-3/4"
            style={{visibility: visibleSourceHandles}}
            type="source"
            position={Position.Bottom}
            id="bottom3"
        />
      </div>

      {
        isConnectContinue && associationType === "through" && (connectionStartId !== id) && 
        <div className="text-tiny text-center text-first-100 bg-first-500  w-16 h-full absolute right-full bottom-0 rounded-l-md px-1 content-center flex items-center " onMouseEnter={(event) => onMouseEnterThrough(event, id)}>
          <div className='text-center w-full'>
            through
          </div>
        </div>
      }
    </div>
  );
}

      