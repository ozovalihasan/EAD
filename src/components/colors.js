const colors = {
  suitable: '#F4FECD',
  disabled: '#CFDBD5',
  association: '#FFF',
  attributeContainer: '#AAFAE9',
  attribute: '#AAFAE9',
  entity: '#94F6EA',
  entityContainer: '#94F6EA',
  EAD: '#FFF',
  factory: '#FFF',
  entityClone: '#F5C987',
  chosen: '#52C9CF',
  warning: '#F94144',
};

export default colors;
