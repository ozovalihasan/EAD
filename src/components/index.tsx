export * from './Markers/CrossMarker'
export * from './Markers/CrowsFootMarker'
export * from './Markers/CircleLineMarker'
export * from './Markers/StraightLineMarker'
export * from './TargetHandle/TargetHandle'
export * from './SelectThroughNode/SelectThroughNode'
export * from './CustomHandle/CustomHandle'
export * from './ConnectionLine/ConnectionLine'
export * from './HasAnyEdge/HasAnyEdge';
export * from './HasManyEdge/HasManyEdge';
export * from './HasOneEdge/HasOneEdge';
export * from './ThroughEdge/ThroughEdge';
export * from './Sidebar/Sidebar';
export * from './Markers/MarkerBase';
export * from './EntityNode/EntityNode';
export * from './HasManyHandle/HasManyHandle';
export * from './HasOneHandle/HasOneHandle';
export * from './TableName/TableName';
export * from './ThroughHandle/ThroughHandle';
export * from './RemoveEdgeButton/RemoveEdgeButton';
export * from './RemoveNodeEdgeButton/RemoveNodeEdgeButton';
export * from './ToggleOptionalButton/ToggleOptionalButton';
export * from './ShowEdgeText/ShowEdgeText';
export * from './Navbar/Navbar';
export * from './EADLogo/EADLogo';
export * from './GithubLogo/GithubLogo';
export * from './AllHandlers/AllHandlers';
export * from './Settings/Settings';
export * from './App/App';
export * from './SidebarHandle/SidebarHandle';
export * from './MainReactFlow/MainReactFlow';
export * from './AttributeTypeOptions/AttributeTypeOptions';
export * from './SidebarOptions/SidebarOptions';
export * from './MainColorDropdown/MainColorDropdown';
export * from './TableOptions/TableOptions';
export * from './TableAttributes/TableAttributes';
export * from './AddTableButton/AddTableButton';
export * from './AddAttributeButton/AddAttributeButton';
export * from './RemoveAttributeButton/RemoveAttributeButton';
export * from './ReorderAllTablesButton/ReorderAllTablesButton';
export * from './ProjectName/ProjectName';
