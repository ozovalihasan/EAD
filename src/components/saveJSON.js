/* eslint-disable no-console */
const saveJSON = (data, filename) => {
  let dataNew = data;
  let filenameNew = filename;
  if (!dataNew) {
    console.error('No data');
    return;
  }

  if (!filenameNew) { filenameNew = 'console.json'; }

  if (typeof dataNew === 'object') {
    dataNew = JSON.stringify(dataNew, undefined, 4);
  }

  const blob = new Blob([dataNew], { type: 'text/json' });
  const e = document.createEvent('MouseEvents');
  const a = document.createElement('a');

  a.download = filenameNew;
  a.href = window.URL.createObjectURL(blob);
  a.dataset.downloadurl = ['text/json', a.download, a.href].join(':');
  e.initMouseEvent('click', true, false, window, 0, 0, 0, 0, 0, false, false, false, false, 0, null);
  a.dispatchEvent(e);
};

export default saveJSON;
