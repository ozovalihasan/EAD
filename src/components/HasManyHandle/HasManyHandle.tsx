import { CustomHandle } from '@/components';

export const HasManyHandle = ({nodeId} : {nodeId: string}) => {

  return (
    <div className="w-6 h-6 absolute -bottom-3 left-1/2 bg-slate-700 rounded-full -translate-x-1/2" >
      <svg 
        className="stroke-[20] stroke-slate-50 w-6 h-6 "
        viewBox="0 0 200 200" 
        xmlns="http://www.w3.org/2000/svg"
      >
        <path 
          style={{strokeLinecap:"round"}} 
          d="M100 20v160 M100 20 20 180 m80-160 80 160"
        />
      </svg>

      <CustomHandle
        handleType='has_many'
        nodeId={nodeId}
        id="bottom2"
      />
    </div>
  )
}
