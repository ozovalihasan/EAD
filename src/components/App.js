import React from 'react';
import { DragDropContext, Droppable } from 'react-beautiful-dnd';
import { useDispatch, useSelector } from 'react-redux';
import { library } from '@fortawesome/fontawesome-svg-core';
import {
  faArrowsAlt, faExpandAlt, faCompressAlt, faEllipsisH, faEllipsisV, faFlag, faClone,
  faPlus, faPlaneDeparture, faPlaneArrival, faTimes,
} from '@fortawesome/free-solid-svg-icons';
import { faGithub } from '@fortawesome/free-brands-svg-icons';
import styled from 'styled-components';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import Model from './Model';
import {
  removeItem,
  addItem,
  moveItem,
  updateDraggedItemId,
  resetState,
  installState,
  idCountIncrease,
  toggleExpandAll,
  toggleCompactMode,
  updateRestrictedDropId,
} from '../redux';
import saveJSON from './saveJSON';
import colors from './colors';

library.add(faArrowsAlt, faExpandAlt, faCompressAlt, faEllipsisH,
  faEllipsisV, faFlag, faClone, faGithub, faPlus, faPlaneDeparture, faPlaneArrival, faTimes);

const App = () => {
  const {
    items, dragDropCategory, idCount, expandAll, compactMode,
  } = useSelector((state) => state.block);

  const startingId = compactMode ? 9 : 0;
  const dispatch = useDispatch();

  const checkDragDropCategory = (dragId, dropId) => (
    dragDropCategory[items[dragId].category].includes(items[dropId].category)
  );

  const stateBlock = useSelector((state) => state.block);

  const saveBlocks = () => {
    localStorage.block = JSON.stringify(stateBlock);
  };

  const onDragStart = (result) => {
    dispatch(updateDraggedItemId(result.draggableId));
  };

  const onDragEnd = (result) => {
    dispatch(updateDraggedItemId(-1));
    const {
      destination, draggableId, source,
    } = result;

    if (!destination) {
      if (!(items[draggableId].factory)) {
        dispatch(removeItem(source.droppableId, source.index, draggableId));
      }
      return;
    }

    if (!checkDragDropCategory(draggableId, destination.droppableId)) {
      return;
    }

    if (items[draggableId].factory) {
      dispatch(addItem(draggableId, destination.droppableId, destination.index, idCount));
      if (items[draggableId].association) {
        dispatch(updateRestrictedDropId(idCount, null));
      }
      dispatch(idCountIncrease());
      return;
    }

    if (items[destination.droppableId].entityContainer
      && items[destination.droppableId].subItemIds.length > 0
      && (items[items[destination.droppableId].subItemIds[0]].category
        !== items[draggableId].category
      )) {
      return;
    }
    dispatch(moveItem(
      draggableId,
      destination.droppableId,
      destination.index,
      source.droppableId,
      source.index,
    ));
  };

  return (
    <MainContainer className="App">
      <LogoButtons>
        <Logo src={`${process.env.PUBLIC_URL}/images/ead-logo.svg`} alt="EAD logo" />
        <Button
          onClick={() => saveJSON(items, 'EAD.json')}
          type="button"
        >
          Download EAD
        </Button>

        <CompactMode
          onClick={() => dispatch(toggleCompactMode())}
          type="button"
          compactMode={compactMode}
        >
          Compact Mode
        </CompactMode>

        {compactMode
        || (
          <>
            <Button
              onClick={() => dispatch(resetState())}
              type="button"
            >
              Reset
            </Button>
            <Button
              onClick={() => saveBlocks()}
              type="button"
            >
              Save
            </Button>
            <Button
              onClick={() => dispatch(installState())}
              type="button"
            >
              Install Saved Data
            </Button>

            <ExpandAllButton
              onClick={() => dispatch(toggleExpandAll())}
              type="button"
              expandAll={expandAll}
            >
              Expand All
            </ExpandAllButton>

            <GithubLink href="https://github.com/ozovalihasan/ead">
              <FontAwesomeIcon icon={faGithub} size="2x" />
            </GithubLink>
          </>
        )}
      </LogoButtons>
      <DragDropContext
        onDragStart={onDragStart}
        onDragEnd={onDragEnd}
      >
        <Droppable
          droppableId="OuterDrop"
          direction="vertical"
          isDropDisabled
        >

          { (provided) => (
            <div
              {...provided.droppableProps}
              ref={provided.innerRef}
            >
              <Model
                parentId={startingId}
                item={items[startingId]}
                allItems={items}
                index={startingId}
                checkDragDropCategory={checkDragDropCategory}
              />
              {provided.placeholder}
            </div>
          )}

        </Droppable>
      </DragDropContext>
    </MainContainer>
  );
};

const MainContainer = styled.div`
  font-family: Arial, Helvetica, sans-serif;
`;

const LogoButtons = styled.div`
  display: flex;
  width: 100%;
  align-items: center;
`;

const Logo = styled.img`
  width: 60px;
  padding-right: 40px;
`;

const Button = styled.button`
  border-radius: 5px;
  border: 1px solid ${colors.disabled};
  background-color: transparent;
  margin: 10px;
  padding: 10px;
  transition: background-color 0.2s;

  &:hover {
    cursor: pointer;
    background-color: ${colors.entity};
  }
`;

const ExpandAllButton = styled(Button)`
  background-color: ${(props) => (props.expandAll && colors.association)};
`;

const CompactMode = styled(Button)`
  background-color: ${(props) => (props.compactMode && colors.entity)};
`;

const GithubLink = styled.a`
  margin: 0 10px;
  color: #000;

  &:hover {
    color: ${colors.chosen};
  }
`;
export default App;
