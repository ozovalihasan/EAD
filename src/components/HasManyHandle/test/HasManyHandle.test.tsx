import { HasManyHandle } from '../HasManyHandle';
import { render, screen } from "@testing-library/react";
import { CustomHandleType } from '@/components';

jest.mock('@/components',  () => ({
  CustomHandle: ({
    nodeId,
    id,
    handleType
  }: CustomHandleType) => {
    
    return (
      <>
        nodeId: {nodeId}
        id: {id}
        handleType: {handleType}
      </>
    );
  },
}));

let renderReadyComponent: JSX.Element;

beforeEach(() => {
  renderReadyComponent = (
    <HasManyHandle nodeId="1" />
  );
});

describe('<HasManyHandle />', () => {
  it('calls another component with a prop with value "through"', () => {
    render(renderReadyComponent);
    expect(screen.getByText(/has_many/i)).toBeInTheDocument();
  });
  

  it('renders correctly', () => {
    const renderedContainer = render(renderReadyComponent );
    expect(renderedContainer).toMatchSnapshot();
  });

});