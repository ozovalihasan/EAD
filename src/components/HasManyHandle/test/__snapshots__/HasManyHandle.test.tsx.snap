// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HasManyHandle /> renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="w-6 h-6 absolute -bottom-3 left-1/2 bg-slate-700 rounded-full -translate-x-1/2"
      >
        <svg
          class="w-6 h-6"
          viewBox="0 0 200 200"
        >
          <path
            class="fill-slate-50"
            d="M 100,0 0,144.9335 V 200 H 17.807516 L 84.697214,66.2206 V 200 H 111.26569 L 110.92704,63.87718 173.53021,200 H 200 v -46.14664 z"
          />
        </svg>
        nodeId: 
        1
        id: 
        bottom2
        handleType: 
        has_many
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="w-6 h-6 absolute -bottom-3 left-1/2 bg-slate-700 rounded-full -translate-x-1/2"
    >
      <svg
        class="w-6 h-6"
        viewBox="0 0 200 200"
      >
        <path
          class="fill-slate-50"
          d="M 100,0 0,144.9335 V 200 H 17.807516 L 84.697214,66.2206 V 200 H 111.26569 L 110.92704,63.87718 173.53021,200 H 200 v -46.14664 z"
        />
      </svg>
      nodeId: 
      1
      id: 
      bottom2
      handleType: 
      has_many
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
